File: ha-git-ops/src/web.go
package main

// Ingress UI: status, per-file diffs, refresh / apply / promote actions.
// Served under HA ingress — all URLs are relative so the ingress path
// prefix is transparent. Secrets diffs are masked to key names only.

import (
	_ "embed"
	"html/template"
	"log"
	"net/http"
	"strings"
)

//go:embed page.gohtml
var pageSrc string

var page = template.Must(template.New("page").Parse(pageSrc))

type fileView struct {
	Rel, Why, Diff string
	CanRevert      bool
}

// done returns a tiny self-refreshing page instead of a 303 — HA's
// ingress iframe mishandles redirect responses to form POSTs and gets
// stuck on a loading screen.
func done(w http.ResponseWriter, what string) {
	w.Header().Set("Content-Type", "text/html; charset=utf-8")
	_, _ = w.Write([]byte(`<!doctype html><meta http-equiv="refresh" content="1;url=./">` +
		`<body style="font-family:system-ui;margin:1.5rem">✓ ` +
		template.HTMLEscapeString(what) + ` — refreshing…</body>`))
}

type pageView struct {
	StatusHTML      template.HTML
	RestartRequired bool
	Conflicts       []fileView
	Incoming        []fileView
	Files           []fileView
	PromoteAll      bool
	InSync          bool
}

func NewWebUI(rec *Reconciler) http.Handler {
	mux := http.NewServeMux()

	mux.HandleFunc("GET /{$}", func(w http.ResponseWriter, req *http.Request) {
		st, sha := rec.Snapshot()
		if len(sha) > 9 {
			sha = sha[:9]
		}
		v := pageView{
			RestartRequired: st.RestartRequired,
			InSync: st.Error == "" && len(st.Conflicts) == 0 &&
				len(st.Drift) == 0 && len(st.Incoming) == 0,
		}

		var status string
		switch {
		case st.Error != "":
			status = `<span class="bad">✗ ` + template.HTMLEscapeString(st.Error) + `</span>`
		case len(st.Conflicts) > 0:
			status = `<span class="bad">⚠ conflicts</span>`
		case len(st.Incoming) > 0:
			status = `<span class="info">↓ updates in git</span>`
		case len(st.Drift) > 0:
			status = `<span class="warn">● drift</span>`
		default:
			status = `<span class="ok">✓ in sync</span>`
		}
		last := st.LastSync
		if last == "" {
			last = "never"
		}
		v.StatusHTML = template.HTML(status +
			` <span class="nw">applied <code>` + template.HTMLEscapeString(sha) + `</code></span>` +
			` <span class="nw">· last refresh ` + template.HTMLEscapeString(last) + `</span>`)

		for _, rel := range sortedKeys(st.Conflicts) {
			v.Conflicts = append(v.Conflicts, fileView{Rel: rel, Why: st.Conflicts[rel]})
		}
		for _, rel := range sortedKeys(st.Incoming) {
			v.Incoming = append(v.Incoming, fileView{
				Rel: rel, Why: st.Incoming[rel], Diff: rec.IncomingDiffText(rel)})
		}
		for _, rel := range sortedKeys(st.Drift) {
			v.Files = append(v.Files, fileView{
				Rel: rel, Why: st.Drift[rel], Diff: rec.DiffText(rel),
				CanRevert: rec.CanRevert(rel)})
		}
		v.PromoteAll = len(v.Files) > 1
		w.Header().Set("Content-Type", "text/html; charset=utf-8")
		_ = page.Execute(w, v)
	})

	mux.HandleFunc("POST /refresh", func(w http.ResponseWriter, req *http.Request) {
		if err := rec.RefreshNow(); err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}
		done(w, "refreshed")
	})

	mux.HandleFunc("POST /apply", func(w http.ResponseWriter, req *http.Request) {
		if err := rec.ApplyUpstream(); err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}
		done(w, "applied updates from git")
	})

	mux.HandleFunc("POST /revert", func(w http.ResponseWriter, req *http.Request) {
		rel := req.FormValue("rel")
		if rel == "" || strings.Contains(rel, "..") {
			http.Error(w, "bad rel", http.StatusBadRequest)
			return
		}
		if err := rec.Revert(rel); err != nil {
			log.Printf("revert %s failed: %v", rel, err)
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}
		done(w, "applied git version of "+rel)
	})

	mux.HandleFunc("POST /promote", func(w http.ResponseWriter, req *http.Request) {
		if err := req.ParseForm(); err != nil {
			http.Error(w, err.Error(), http.StatusBadRequest)
			return
		}
		rels := req.PostForm["rel"]
		if len(rels) == 0 {
			http.Error(w, "no files", http.StatusBadRequest)
			return
		}
		for _, rel := range rels {
			if strings.Contains(rel, "..") {
				http.Error(w, "bad rel", http.StatusBadRequest)
				return
			}
		}
		if err := rec.Promote(rels, req.FormValue("message")); err != nil {
			log.Printf("promote %v failed: %v", rels, err)
			http.Error(w, err.Error(), http.StatusConflict)
			return
		}
		done(w, "promoted")
	})

	return mux
}

File: ha-git-ops/rootfs/opt/ha_gitops/web.py
"""Ingress UI: status, per-file diffs, promote / revert / sync actions.

Served under HA ingress — all URLs are relative so the ingress path
prefix is transparent. Secrets diffs are masked to key names only.
"""

import difflib
import html
import logging

from aiohttp import web

from reconciler import SECRETS_PLAIN

log = logging.getLogger("web")

PAGE = """<!doctype html><html><head><meta charset="utf-8">
<title>HA GitOps</title><style>
 body {{ font-family: system-ui, sans-serif; margin: 1.5rem; max-width: 60rem; }}
 .ok {{ color: #1a7f37; }} .bad {{ color: #cf222e; }} .warn {{ color: #9a6700; }}
 pre {{ background: #f6f8fa; padding: .8rem; overflow-x: auto; font-size: .85rem; }}
 pre .add {{ color: #1a7f37; }} pre .del {{ color: #cf222e; }}
 .file {{ border: 1px solid #d0d7de; border-radius: 6px; margin: 1rem 0; }}
 .file > header {{ padding: .5rem .8rem; background: #f6f8fa;
   display: flex; justify-content: space-between; align-items: center; }}
 button {{ cursor: pointer; }} input[type=text] {{ width: 24rem; }}
</style></head><body>
<h2>HA GitOps</h2>
<p>{status}</p>
<form method="post" action="./sync" style="display:inline">
  <button>Sync now</button></form>
{restart}
{conflicts}
{files}
{promote_all}
</body></html>"""


def _diff_html(rel, want, live):
    if rel == SECRETS_PLAIN:
        return "<pre>(secrets diff masked — values changed)</pre>"
    want_l = (want or b"").decode(errors="replace").splitlines()
    live_l = (live or b"").decode(errors="replace").splitlines()
    lines = []
    for l in difflib.unified_diff(want_l, live_l, "git (desired)",
                                  "live", lineterm=""):
        cls = "add" if l.startswith("+") else "del" if l.startswith("-") else ""
        lines.append(f'<span class="{cls}">{html.escape(l)}</span>')
    return "<pre>" + "\n".join(lines) + "</pre>"


def make_app(rec):
    async def index(_):
        st = rec.status
        drift, conflicts = st["drift"], st["conflicts"]
        if st.get("error"):
            status = f'<span class="bad">✗ {html.escape(st["error"])}</span>'
        elif conflicts:
            status = f'<span class="bad">⚠ {len(conflicts)} conflict(s)</span>'
        elif drift:
            status = f'<span class="warn">● {len(drift)} file(s) drifted</span>'
        else:
            status = '<span class="ok">✓ in sync</span>'
        status += (f' — applied <code>{(rec.state["applied_sha"] or "")[:9]}'
                   f'</code>, last sync {st["last_sync"] or "never"}')

        restart = ""
        if st["restart_required"]:
            restart = ('<p class="warn">Core restart required to activate '
                       'applied changes.</p>')

        conflicts_html = ""
        if conflicts:
            items = "".join(f"<li><code>{html.escape(r)}</code> — "
                            f"{html.escape(why)}</li>"
                            for r, why in sorted(conflicts.items()))
            conflicts_html = (f"<h3>Conflicts (resolve by revert or promote, "
                              f"then sync)</h3><ul>{items}</ul>")

        blocks = []
        for rel in sorted(drift):
            want = rec._at_ref(rec.state["applied_sha"], rel)
            live = rec._live(rel)
            blocks.append(f"""<div class="file"><header>
  <code>{html.escape(rel)}</code> <span>{html.escape(drift[rel])}</span>
  <span>
   <form method="post" action="./revert" style="display:inline">
    <input type="hidden" name="rel" value="{html.escape(rel)}">
    <button>Revert to git</button></form>
   <form method="post" action="./promote" style="display:inline">
    <input type="hidden" name="rel" value="{html.escape(rel)}">
    <input type="text" name="message" placeholder="commit message">
    <button>Promote</button></form>
  </span></header>
  {_diff_html(rel, want, live)}</div>""")

        promote_all = ""
        if len(drift) > 1:
            promote_all = f"""<form method="post" action="./promote">
  {"".join(f'<input type="hidden" name="rel" value="{html.escape(r)}">'
           for r in sorted(drift))}
  <input type="text" name="message" placeholder="commit message">
  <button>Promote all {len(drift)} files</button></form>"""

        return web.Response(
            text=PAGE.format(status=status, restart=restart,
                             conflicts=conflicts_html,
                             files="".join(blocks), promote_all=promote_all),
            content_type="text/html")

    async def sync(_):
        await rec.tick()
        raise web.HTTPFound("./")

    async def revert(request):
        form = await request.post()
        await rec.revert(form["rel"])
        raise web.HTTPFound("./")

    async def promote(request):
        form = await request.post()
        rels = form.getall("rel")
        try:
            await rec.promote(rels, form.get("message", ""))
        except RuntimeError as e:
            return web.Response(text=str(e), status=409)
        raise web.HTTPFound("./")

    app = web.Application()
    app.router.add_get("/", index)
    app.router.add_post("/sync", sync)
    app.router.add_post("/revert", revert)
    app.router.add_post("/promote", promote)
    return app
